src/workerd/api/node/zlib-util.c++

```diff
@@ -878,6 +838,129 @@ void ZlibUtil::CompressionStream<CompressionContext>::FreeForZlib(void* data, vo
   auto real_pointer = static_cast<uint8_t*>(pointer) - sizeof(size_t);
   JSG_REQUIRE(ctx->allocations.erase(real_pointer), Error, "Zlib allocation should exist"_kj);
 }
+namespace {
+// A RAII wrapper around a compression context class
+template <typename CompressionContext>
+class ContextRAII: public CompressionContext {
+public:
+  using CompressionContext::CompressionContext;
+
+  ~ContextRAII() {
+    static_cast<CompressionContext*>(this)->close();
+  }
+};
+
+template <typename Context>
+static kj::Array<kj::byte> syncProcessBuffer(Context& ctx, GrowableBuffer& result) {
+  do {
+    result.addChunk();
+    ctx.setOutputBuffer(kj::ArrayPtr(result.end(), result.available()));
+
+    ctx.work();
+
+    KJ_IF_SOME(error, ctx.getError()) {
+      JSG_FAIL_REQUIRE(Error, error.message);
+    }
+
+    result.adjustUnused(ctx.getAvailOut());
+  } while (ctx.getAvailOut() == 0);
+
+  return result.releaseAsArray();
```

Comment:
shouldn't we need attach here, in order to prevent a memory leak? @jasnell

```suggestion
  return result.releaseAsArray().attach(kj::mv(result));
```

Reply:
No, that's not needed in this case. `releaseAsArray()` is going to return a `kj::Array<T>` that owns the data, basically moving the underlying storage away from `result`. You'd be uselessly attaching an empty object here.